AmethystAPI/src/amethyst/runtime/patchManager/PatchManager.cpp

```diff
@@ -0,0 +1,66 @@
+#include "PatchManager.h"
+#include "../../Log.h"
+#include "../../Memory.h"
+namespace Amethyst {
+    // Returns false if the patch failed to apply
+    bool PatchManager::ApplyPatch(uintptr_t address, uint8_t* patch, size_t size){
+        DWORD oldProtection;
+        if (!UnprotectMemory(address, size, &oldProtection)) {
+            Assert("Failed to unprotect memory at address: 0x{0:x}", address);
+			return false;
+		}
+
+        uint64_t boundsLow = address;
+        uint64_t boundsHigh = address + size;
+        for (auto& establishedPatch : this->mPatches) {
+			if (establishedPatch.address == address) {
+                Log::Warning("Patch already exists at address: 0x{0:x}", address);
+			}
+            uint64_t establishedBoundsLow = establishedPatch.address;
+            uint64_t establishedBoundsHigh = establishedPatch.address + establishedPatch.original.size;
+            if (boundsLow >= establishedBoundsLow && boundsLow <= establishedBoundsHigh
+                || boundsHigh >= establishedBoundsLow && boundsHigh <= establishedBoundsHigh) {
+                Log::Warning("Patch at address: 0x{0:x} overlaps with existing patch at address: 0x{1:x}", address, establishedPatch.address);
+			}
+		}
+        Patch ptch;
+		OriginalMemory original;
+		uint8_t* originalMemory = (uint8_t*)malloc(size);
+        if (!originalMemory) 
+			Assert("Failed to allocate memory for original bytes at address: 0x{0:x}", address);
+        memcpy(originalMemory, reinterpret_cast<void*>(address), size);
+		original.original = originalMemory;
+        original.size = size;
+		ptch.original = original;
+        ptch.address = address;
+        memcpy(reinterpret_cast<void*>(address), patch, size);
+        this->mPatches.emplace_back(ptch);
+        if (!ProtectMemory(address, size, oldProtection)) {
+            Assert("Failed to protect memory at address: 0x{0:x}", address);
```

Comment:
If we assert we shouldn't need to return false, plus the assert should be being thrown by ProtectMemory. I'm open for a form of this function where it doesn't throw (Like SigScanSafe), but the default behaviour if it fails should be to assert

Reply:
Ah yeah i do agree. That was left over from an older version. Although i dont think ProtectMemory should throw

Reply:
I would rather a mod throw an exception because an operation failed earlier than later. This could lead to some confusing behaviours if it fails to apply the patches, and if it does fail to apply the patches the mod won't function etc...

Plus if the operation fails its more of an indication shit went wrong because there's not really a reason for it to have failed

Reply:
True, even if it is on the mod developer for not checking it.  
Ill switch that to an assert in the function